### DISCUSSION TITLE
Host docusaurus source in Azure DevOps

### DISCUSSION THREAD

**Discussion by gentledepp** [facebook/docusaurus#7171] [Q&A]
Hi!

While docusaurus supports github and gitlab, I would like to use Azure DevOps to manage our documentation repository.
Since Azure DevOps also runs git and allows for editing files in the browser, this should be doable, I think.

Could you please point me to the right direction, so I can figure out how to provide a PR for this?`

The only integration to github I can see is
- [x] for each markdown page, there is an "edit" button that brings you directly to the source code on github.
- [x] you can use git branches on github to actually version the documentation

Is there anything I am missing?

Thanks for your valuable insights!

**Comment by gentledepp** [comment]
Hi! 
So it seems the configuring this in the docusaurus.config.js is easy:
1) repository URL
2) edit link
![image](https://user-images.githubusercontent.com/1175751/163385752-54154cea-9785-471f-8c92-a37eaab65ac2.png)

for (1), you need to change this link:
![image](https://user-images.githubusercontent.com/1175751/163386200-89abad2a-8ee9-42df-b9e5-10ab4fc9a34a.png)

for (2), you simply have to change the `editUrl`
This also seems to be fine, since, if a document is in "docs/designer/designer.md", this path will just be appended to the `editUrl` configured in the docusaurus.config.js

To get the link, browse to one of the markdown files in your docusaurus repository. Then copy the URL, but remove the value of the "path=" URL parameter since it will be replaced:
![image](https://user-images.githubusercontent.com/1175751/163386698-11b57d0f-e779-4b98-8b1d-9763151c5a56.png)

Done.

However, as I understood correctly, docusaurus also supports having multiple versions of a documentation using git branches.
I currently do not see how this would work with Azure DevOps.

Does anyone have some insights on this?

**Reply by slorber** [reply]
> However, as I understood correctly, docusaurus also supports having multiple versions of a documentation using git branches.

By default, Docusaurus use multiple versions on the same Git branch, see for example our own website: https://github.com/facebook/docusaurus/tree/main/website/versioned_docs

Now it's up to you to use one Git branch for each version. This also works, but you'll likely have to set up a separate CDN (Azure DevOps) deployment/subdomain per git-branch/version.

Remember Docusaurus responsibility is only to build a static site (produce HTML files). Now it's up to you to figure out what's the best way to expose those files to the world. We do recommend and document Netlify/Vercel/GH more extensively, but any other host can work.

**Reply by gentledepp** [reply]
Aaah I see! So docusaurus does not access git at runtime... my bad. I got confused.
This makes much more sense now.
So I assume I will just have to look at the Netlify/Vercel/GH samples and modify them to work with azure. No problem. Thank you so much!

**Comment by wagnbeu0** [comment]
Hi, one question from my side:

I have a Azure DevOps Repo running at https://dev.azure.com/MyCompany/MyProject/_git/MyRepo

I configured the docusaurus.config.js to:
`
 docs: {
 routeBasePath: '/',
 sidebarPath: './sidebars.js',
 // Please change this to your repo.
 // Remove this to remove the "edit this page" links.
 editUrl:
 'https://dev.azure.com/MyCompany/MyProject/_git/MyRepo',
`

My repo is cloned local to `/opt/ARTOPS_DOKU/docs/Docu`, Docusaurus is installed under `/opt/ARTOPS_DOKU`,

When I now click on the edit button, i get an error message that the file does not exist:
`https://dev.azure.com/MyCompany/MyProject/_git/MyRepo/docs/Docu/documentation/Testfile.md`

But the file is located in the repo under `https://dev.azure.com/MyCompany/MyProject/_git/MyRepo/documentation/Testfile.md`

So how should I change the parameters?

**Reply by slorber** [reply]
@wagnbeu0 I don't know how Azure works so I can't answer what should be the edit url target.

But you can construct the edit url yourself thanks to plugin options, so once you figure out what the url should be, you should be able to implement a solution.

https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-docs#EditUrlFunction

**Comment by wagnbeu0** [comment]
Hm, I see this EditURLFunctions, but have no clue how to remove stuff from the URL. Here the docu is not clear to me. In the docusaurus docu I found the following lines:
The final URL is computed by editUrl + relativeDocPath.

So I need a way to remove the relaticeDocPath which is obsolete.

**Reply by slorber** [reply]
This final url is only when you use the string form of that url.

If you do: 

```js
editUrl: (params) => `https://azure.com`,
```

Then all your docs will target that hardcoded value. You have complete freedom to turn the appropriate edit url from each doc, using the params to create that url. We will use as a target url exactly what you return.
